File: Cpp/LeetCode/146.py
class ListNode:
    def __init__(self, v, succ=None, prev=None):
        self.v = v
        self.succ = succ
        self.prev = prev

class List:
    def __init__(self):
        self.dummy = ListNode(-1)
        self.tail = self.dummy

    def append(self, v):
        self.tail.succ = ListNode(v, None, self.tail)
        self.tail = self.tail.succ
    
    def pophead(self):
        return self.remove(self.dummy.succ)
    
    def remove(self, p):
        p.prev.succ = p.succ
        p.succ.prev = p.prev
        return p.v

class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.l = List()
        self.d = {}

    def get(self, key: int) -> int:
        if key in self.d:
            _, value = self.l.remove(self.d[key])
            self.l.append((key, value))
            return value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.d:
            self.l.remove(self.d[key])
        else:
            if len(self.d) >= self.cap:
                k, v = self.l.pophead()
                self.d.pop(k)
        self.l.append((key, value))
        self.d[key] = self.l.tail


# Your LRUCache object will be instantiated and called as such:
obj = LRUCache(1)
obj.put(2, 1)
param_1 = obj.get(2)

File: Cpp/LeetCode/146.cpp
#include "usual.h"

class List_Node {
public:
    int key;
    List_Node* succ;
    List_Node* prev;
    
    List_Node(): key(0), succ(nullptr), prev(nullptr) {}
    List_Node(int _key, List_Node* _succ, List_Node* _prev): key(_key), succ(_succ), prev(_prev) {}
    
    void insertAfter(List_Node* ln) {
        if (!ln)
            return;
        ln->succ = succ;
        ln->prev = this;
        if (succ) {
            succ->prev = ln;
        }
        succ = ln;
    }

    void removeAfter() {
        if (!succ)
            return;
        List_Node* tmp = succ;
        if (succ->succ)
            succ->succ->prev = this;
        succ = succ->succ;
        free(tmp);
    }

    void remove() {
        if (prev)
            prev->succ = succ;
        if (succ)
            succ->prev = prev;
    }
};

// class List {
//     List_Node *dummy;
//     List_Node *tail;

// public:
//     List(): dummy(new List_Node()), tail(dummy) {}

// };

class LRUCache {
    int cnt;
    int cap;
    unordered_map<int, pair<int, List_Node*>> cache;
    List_Node* dummy;
    List_Node* tail;

public:
    LRUCache(int capacity): cnt(0), cap(capacity), dummy(new List_Node), tail(dummy) {
        
    }
    
    int get(int key) {
        if (!cache.count(key))
            return -1;
        List_Node *p = cache[key].second;
        if (p != tail) {
            p->remove();
            tail->insertAfter(p);
            tail = tail->succ;
        }
        cout << "After get(" << key << ")" << endl;
        print();
        return cache[key].first;
    }
    
    void put(int key, int value) {
        if (cache.count(key)) {
            if (cache[key].second == tail)
                tail = tail->prev;
            cache[key].second->remove();
            delete cache[key].second;
        } else if (cache.size() >= cap) {
            cache.erase(dummy->succ->key);
            if (dummy->succ == tail)
                tail = tail->prev;
            auto p = dummy->succ;
            dummy->succ->remove();
            delete p;
        }
        tail->succ = new List_Node(key, tail->succ, tail);
        tail = tail->succ;
        cache[key].first = value;
        cache[key].second = tail;
        cout << "After put(" << key << ", " << value << ")" << endl;
        print();
    }

    void print() {
        auto p = dummy->succ;
        while (p) {
            cout << p->key << " ";
            p = p->succ;
        }    
        cout << endl;
        for (const auto& [k, v] : cache)
            cout << "(" << k << ", " << v.first << ") ";
        cout << endl;
    }
};

int main(int argc, char const *argv[])
{
    LRUCache* obj = new LRUCache(2);
    obj->get(2);
    obj->put(2, 6);
    obj->get(1);
    obj->put(1, 5);
    obj->put(1, 2);
    obj->get(1);
    obj->get(2);
    return 0;
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */